ARROW-6216: [C++] Expose codec compression level to user

martinradev opened the pull request:
This patch allows the user to select a compression level to be used for when column data in Parquet is compressed.
The user can select between better compression ratio or speed.

emkornfield commented:
Thank you for the patch.  Please make sure to use "make format" to fix the lint errors.

emkornfield commented on cpp/src/arrow/util/compression_zstd.cc:
```diff
@@ -180,12 +189,27 @@ Status ZSTDCompressor::End(int64_t output_len, uint8_t* output, int64_t* bytes_w
 // ZSTD codec implementation
 
+ZSTDCodec::ZSTDCodec() : Codec(), zstdCompressionLevel_(kZSTDDefaultCompressionLevel) {
+  const char* envZstdCompressionLevel = getenv("ARROW_ZSTD_COMPRESSION_LEVEL");
```
I think this is probably the wrong level to be reading environment variables (it seems like we potentially want this at different levels in the same binary), can it be pushed up the stack so it can perhaps be set explicitly.

pitrou replied:
Also please use the `GetEnvVar` helper function from `arrow/util/io-util.h`.

pitrou replied:
The use case seems to be Parquet file writing, btw, so perhaps the setting should be Parquet-specific?

hatemhelal replied:
I would prefer having this setting added to the `parquet::ColumnProperties` object as that already manages related properties.  I don't really know enough about whether there can be one compression level or if there should be one specific for each codec.

martinradev replied:
Done

emkornfield commented on cpp/src/arrow/util/compression_zstd.cc:
```diff
@@ -180,12 +189,27 @@ Status ZSTDCompressor::End(int64_t output_len, uint8_t* output, int64_t* bytes_w
 
+ZSTDCodec::ZSTDCodec() : Codec(), zstdCompressionLevel_(kZSTDDefaultCompressionLevel) {
+  const char* envZstdCompressionLevel = getenv("ARROW_ZSTD_COMPRESSION_LEVEL");
+  if (envZstdCompressionLevel != NULLPTR) {
```
nullptr is fine in .cc files.

hatemhelal replied:
I didn't know this, does that mean that `nullptr` shouldn't be used in .h files?

Ingvar-Y commented:
It would be nice to also expose compression level for other codecs - gzip, brotli, lz4

hatemhelal commented:
Perhaps this could be added to the `parquet::ColumnProperties`? Your client is still free to set a default value on a project level or even override it on a file-by-file basis.

hatemhelal commented on cpp/src/arrow/util/compression_zstd.cc:
```diff
@@ -180,12 +189,27 @@ Status ZSTDCompressor::End(int64_t output_len, uint8_t* output, int64_t* bytes_w
+    constexpr long int intMax = static_cast<long int>(std::numeric_limits<int>::max());
+    const bool isOutOfIntBounds = ((userSpecifiedLevel < intMin) || (userSpecifiedLevel > intMax));
+    if (hasErrorInParsing == false && isOutOfIntBounds == false) {
+      zstdCompressionLevel_ = static_cast<int>(userSpecifiedLevel);
```
Could you explain the various `static_cast`s being used here?  It looks like the compression level is destined to be an `int` type so I think you can do away with the up-cast then downcast like this:

```cpp
    constexpr int intMin = std::numeric_limits<int>::min());
    constexpr int intMax = std::numeric_limits<int>::max());
    const bool isOutOfIntBounds = ((userSpecifiedLevel < intMin) || (userSpecifiedLevel > intMax));
    if (hasErrorInParsing == false && isOutOfIntBounds == false) {
      zstdCompressionLevel_ = userSpecifiedLevel;
``` 

Not sure if I missed something though?

martinradev replied:
No longer relevant since that patch has changed.

martinradev commented:
Some notes regarding the design decision in the patch:
- As suggested, the compression level becomes part of ColumnProperties and has to be piped to the creation of the codec.
- I explicitly did not add validation for what would be a reasonable compression level. That is because it would add maintenance overhead once the version of some codec is updated. As in, the range of 0, 30 makes sense for version 2.4 but for 2.5 the range is 0, 40. I am just sure that this will be missed when a compressor is updated. So, it's up to the user for them to know what to use, what compressor version is used and what compression level makes sense for his or her needs.
- I added and updated tests to exercise the API but they are really smoke tests. It only verifies that the API is callable and no failures are observed.

martinradev commented:
Alright, I see that there are build failures for clients of Arrow. I can go and try to change my patch to avoid breaking everything.

martinradev commented:
@hatemhelal @emkornfield @Ingvar-Y @pitrou 
Thank you for the initial round of reviews. The code has significantly changed.
Can you have another look?

pitrou commented on cpp/src/arrow/io/compressed-test.cc:
```diff
@@ -152,7 +152,10 @@ class CompressedInputStreamTest : public ::testing::TestWithParam<Compression::t
 
   std::unique_ptr<Codec> MakeCodec() {
     std::unique_ptr<Codec> codec;
-    ABORT_NOT_OK(Codec::Create(GetCompression(), &codec));
```
Why not keep using the 2-argument form?

pitrou commented on cpp/src/arrow/util/compression.h:
```diff
@@ -33,6 +33,28 @@ struct Compression {
 
 namespace util {
 
+inline int32_t GetCompressionCodecDefaultCompressionLevel(Compression::type t) {
```
There doesn't seem to be any point in exposing this function in a header file, is there?

martinradev replied:
Thanks, I moved it to the source file.

pitrou commented on cpp/src/arrow/util/compression_brotli.cc:
```diff
@@ -31,10 +31,6 @@
 
-// Brotli compression quality is max (11) by default, which is slow.
-// We use 8 as a default as it is the best trade-off for Parquet workload.
-constexpr int kBrotliDefaultCompressionLevel = 8;
```
IMHO it was fine to leave this default here.

martinradev replied:
I would like to not do that and keep all of the levels in one place.
Also, some codec source files are not always compiled into the project and thus there would have been an extra ifdef necessary just because of that in the function for retrieving the default compression level for the given compression type.

pitrou commented on cpp/src/parquet/column_writer-test.cc:
```diff
@@ -40,6 +40,23 @@ using schema::PrimitiveNode;
+    GetCompressionCodecDefaultCompressionLevel(Compression::LZO);
+const int32_t kLZ4DDefaultCompressionLevel =
+    GetCompressionCodecDefaultCompressionLevel(Compression::LZ4);
+const int32_t kUncompressedDefaultCompressionLevel =
```
What's the point of all this? It seems this is gratuitously complicating the code, no?

martinradev replied:
The point was to avoid having to write GetCompressionCodecDefaultCompressionLevel(...) in many places.
You might argue that than we should set a default value for the compression level parameter. However, I prefer it to have to be explicitly set within Arrow/Parquet while making it optional for the user. The way I see it this avoids potential improper calls to functions which have multiple default parameters, AND avoids having many functions be duplicated whenever it's not possible to have a default value.
Anyway, I will remove these constants.

pitrou replied:
My point is that the default values are known in Arrow somewhere and there's no point in duplicating them in Parquet. Let Arrow fill in the default if no specific value is passed.

pitrou commented on cpp/src/parquet/types.cc:
```diff
@@ -259,6 +266,34 @@ std::string ConvertedTypeToString(ConvertedType::type t) {
   }
 }
 
+int32_t GetCompressionCodecDefaultCompressionLevel(Compression::type t) {
```
Why is this necessary? Why not use the 2-argument form of `Codec::Create` is the compression level isn't specified?

martinradev replied:
In PageWriter::Open I would like to have to the compression level already selected since there are two different paths selected:: BufferedPageWriter and SerializedPageWriter.
I do not want to introduce any default parameters and leave deduction of the compression level at a level close to the arrow::util::Codec::Create because this means there will be two similar paths maintained - one with the compression level selected and the one with no compression level.
Selecting the compression level as close to the parquet api as possible means that there will be for the most part a single path maintained. However, there is a complication that arrow maintains its own compression enum with an extra bz2 entry. Thus, there's a need for some translation like it was already done and a bit simplified in the newest iteration of the change.
I still followed your advice to some extent which also made the change slightly smaller.

hatemhelal replied:
I think it makes more sense for each codec to "know" its own default compression level rather than localizing/externalizing this information into a function that deals with all the codecs.  From the parquet APIs, I think you could use placeholder value like -1 to signify using the default compression level.

martinradev replied:
No, -1 cannot be a placeholder because it can be a valid compression level for some codec which can be different from the default in Arrow. That's why it's best to query the default compression level close to the API. We cannot have a magic value which signals to use the default value which is some other magic value.

pitrou replied:
> That's why it's best to query the default compression level close to the API.

That doesn't follow. You could e.g. have both `ZSTDCodec()` and `ZSTDCodec(int compression_level)` constructors.

martinradev replied:
A stack trace for creating ZSTDCodec which shows one possible path:
arrow::util::ZSTDCodec::ZSTDCodec
arrow::util::Codec::Create
parquet::GetCodecFromArrow
parquet::SerializedPageWriter::SerializedPageWriter
parquet::PageWriter::Open
parquet::RowGroupSerializer::NextColumn

How should be determined in arrow::util::Codec::Create which constructor to use? Once you answer that, please tell me what should happen parquet::GetCodecFromArrow. Then please do that for the next levels.
Is it somehow possible to select which constructor to use without duplicating paths?

A simpler solution is to determine the default compression level for the specified codec higher up in the stack, for example in parquet::RowGroupSerializer::NextColumn. Then we can make the compression_level explicit, as it is done in my patch, and there will be no need for duplicating code.

I can still add both constructors which will simplify a bit some of the code in a few tests which doesn't bring much benefit.

I would like to hear what you think of all of this.

pitrou replied:
Another possibility is `ZSTDCodec(int compression_level = kDefaultCompressionLevel)` where `kDefaultCompressionLevel` is chosen such that it is unlikely to be confused with an actual compression level (for example `INT_MIN`).

martinradev replied:
`ZSTDCodec(int compression_level = kDefaultCompressionLevel)` this is not any different from your early proposal of having two constructors. On higher levels in the stack it would still have to be determined whether to pass an argument or not.

I do not want to propagate some random value through the whole stack with the hope that it doesn't conflict with what the user might specify.

So, let's see. 20 users decide that they want to pass INT_MIN as a compression level. Sure, it doesn't make sense but it works! ZSTD would be set up to use a compression level of 1 since INT_MIN is this magic value which is a hint that arrow is free to select whatever it wants.
Then, arrow changes the implementation a bit so that instead of INT_MIN we pick the much nicer 0x31337 magic value.
All user code suddenly starts behaving in a different way.

pitrou replied:
> On higher levels in the stack it would still have to be determined whether to pass an argument or not.

No, you could simply pass `kDefaultCompressionLevel`.

> So, let's see. 20 users decide that they want to pass INT_MIN as a compression level. Sure, it doesn't make sense but it works!

I am not very interested in discussing contrived scenarios.

martinradev replied:
> No, you could simply pass kDefaultCompressionLevel.
Yes, but kDefaultCompressionLevel will be that magic value you suggest introducing.

> I am not very interested in discussing contrived scenarios.
It is a hack to have an internal representation which can conflict with what the user provides. Especially, if it can easily be avoided.

hatemhelal replied:
> No, -1 cannot be a placeholder because it can be a valid compression level for some codec which can be different from the default in Arrow.

Ok, thanks for correcting me.  Another suggestion would be to add a method on `ColumnProperties` that can tell you whether to use the codec defaults or override it with a user-supplied value.  Something like:

`bool using_compression_level`

that only returns `true` when a user has manually set the compression level.  I've seen problems like this solved with `boost::optional` but feels like overkill for this case, but as you've seen I'm often wrong so maybe check that out too.

martinradev commented:
Thanks for the reviews.
I had to rebase and I simplified the change a little bit.
@pitrou Can you please add whomever you think is an appropriate reviewer for the changes to the parquet reader/writer implementation?

martinradev commented:
@emkornfield are you comfortable with the changes?

hatemhelal commented on cpp/src/parquet/types.h:
```diff
@@ -668,6 +672,8 @@ PARQUET_EXPORT std::string TypeToString(Type::type t);
 PARQUET_EXPORT std::string FormatStatValue(Type::type parquet_type,
                                            const std::string& val);
 
+PARQUET_EXPORT int32_t GetCompressionCodecDefaultCompressionLevel(Compression::type t);
```
Is there a reason to switch to `int32_t`?  It looks like `int` was previously used.

martinradev replied:
I think I used int32_t from the start for everywhere I propagated the compression level. I also used int32_t since it looks like that's the preferred type in Arrow and other programming guidelines.
I know that the compression libraries used in Arrow have the compression level as 'int'. I don't think this is a problem for any system using Arrow but I could replace all int32_t with int if you expect int and int32_t to not have the same size.

martinradev commented:
@wesm can you please review this patch?

emkornfield commented:
My drive by comment is I tend to agree with @pitrou and @hatemhelal but understand being careful about collisions.  A few suggestions which may or may not be a good idea:
- use boost::optional which is already used in gandiva (preferably we would have a censored version)
- simulate optional in this case by grouping compression type and compression into a class that tracks if compression level is set.  I think this makes for a cleaner API anyway
- I'm ok risking using a sentinel value that might tie our hands for future compression type e.g int_min

Also it would be nice to keep existing APIs in place if at all possible and possibly deprecate them instead of replacing them.

pitrou commented:
If we want to gold-plate this, we should probably make the additional argument a `CompressionOptions` struct rather than a compression level integer.

But I still think a compression level with a default value like `int kDefaultCompressionLevel = INT_MIN` is completely fine. I don't care if the user passes such an unlikely value explicitly.

And using `boost::optional` for this sounds ridiculous.

wesm commented on cpp/src/arrow/util/compression_test.cc:
```diff
@@ -341,19 +335,63 @@ class CodecTest : public ::testing::TestWithParam<Compression::type> {
+    case Compression::UNCOMPRESSED:
+    case Compression::SNAPPY:
+      // Compression level cannot be specified for these
+      // compression types.
```
Does it make sense to return an error if you specify a compression level for a compressor that does not support it? Either way we need tests asserting what the behavior is

martinradev replied:
I decided to mention in the api for parquet-cpp that parquet/arrow do not do validation on the passed compression level.
1. it could be some effort to track what is the range of possible compression levels for a version of a compressor. Thus, it's best if the user finds out what's meaningful. Worst case initializing the codec would fail.
2. With that I also decided that I would not add validation for whether a compressor supports specifying a compression level. It's a hint and it doesn't necessarily lead to better compression ratio or speed.

I hope you're fine with that.

wesm commented on cpp/src/arrow/util/compression_zlib.cc:
```diff
@@ -34,8 +34,6 @@
 namespace arrow {
 namespace util {
 
-constexpr int kGZipDefaultCompressionLevel = 9;
```
Same

martinradev replied:
Undid my change with regards to this.

wesm commented on cpp/src/parquet/column_io_benchmark.cc:
```diff
@@ -38,8 +38,9 @@ std::shared_ptr<Int64Writer> BuildWriter(int64_t output_size,
-      PageWriter::Open(dst, Compression::UNCOMPRESSED, metadata);
+  std::unique_ptr<PageWriter> pager = PageWriter::Open(
+      dst, Compression::UNCOMPRESSED,
+      GetCompressionCodecDefaultCompressionLevel(Compression::UNCOMPRESSED), metadata);
```
can this just be `/*compression_level=*/-1`?

wesm commented on cpp/src/parquet/column_writer_test.cc:
```diff
@@ -106,9 +106,11 @@ class TestPrimitiveWriter : public PrimitiveTypedTest<TestType> {
-    std::unique_ptr<PageWriter> pager =
-        PageWriter::Open(sink_, column_properties.compression(), metadata_.get());
+    std::unique_ptr<PageWriter> pager = PageWriter::Open(
+        sink_, compression, GetCompressionCodecDefaultCompressionLevel(compression),
```
Do you want to get the compression level from `column_properties`?

wesm commented on cpp/src/parquet/column_writer_test.cc:
```diff
@@ -397,61 +404,87 @@ TYPED_TEST(TestPrimitiveWriter, RequiredRLEDictionary) {
-                                 LARGE_SIZE);
+  this->TestRequiredWithSettings(
+      Encoding::PLAIN, Compression::SNAPPY, false, false,
+      GetCompressionCodecDefaultCompressionLevel(Compression::SNAPPY), LARGE_SIZE);
```
Can you introduce something to avoid the duplicated instances of `GetCompressionCodecDefaultCompressionLevel(codec)`? This code diff (with this function used 12 times) suggests a needed refactor. I would suggest a struct of some kind to pass into the test function

wesm commented on cpp/src/parquet/file_writer.cc:
```diff
@@ -27,15 +27,33 @@
+                                       Compression::type& outCompressionCodec,
+                                       int32_t& outCompressionLevel) {
+  const Compression::type compression = properties->compression(path);
+  const bool isCompressionLevelSelected = properties->is_compression_level_selected(path);
```
We use `snake_case` for variable names in this project

wesm commented on cpp/src/parquet/file_writer.cc:
```diff
@@ -27,15 +27,33 @@
+void DetermineCompressionCodecAndLevel(const WriterProperties* properties,
+                                       const std::shared_ptr<schema::ColumnPath>& path,
+                                       Compression::type& outCompressionCodec,
+                                       int32_t& outCompressionLevel) {
```
Does not conform to style guide

* Use pointers for "out" parameters
* Use `snake_case` for variable names

wesm commented on cpp/src/parquet/file_writer.cc:
```diff
@@ -27,15 +27,33 @@
 
+namespace {
+
+void DetermineCompressionCodecAndLevel(const WriterProperties* properties,
```
Seems like this should be a method on `WriterProperties`

wesm commented on cpp/src/parquet/file_writer.cc:
```diff
@@ -124,10 +142,12 @@ class RowGroupSerializer : public RowGroupWriter::Contents {
-    std::unique_ptr<PageWriter> pager =
-        PageWriter::Open(sink_, properties_->compression(column_descr->path()), col_meta,
-                         properties_->memory_pool());
+    Compression::type compressionCodec;
```
Non-conforming name

wesm commented on cpp/src/parquet/file_writer.cc:
```diff
@@ -124,10 +142,12 @@ class RowGroupSerializer : public RowGroupWriter::Contents {
+    DetermineCompressionCodecAndLevel(properties_, col_meta->descr()->path(),
+                                      compressionCodec, compression_level);
+    std::unique_ptr<PageWriter> pager = PageWriter::Open(
+        sink_, compressionCodec, compression_level, col_meta, properties_->memory_pool());
```
This almost suggests it would be better to pass `properties_` and the column path into `PageWriter::Open` instead. Not sure how much refactoring that would be

martinradev replied:
I think it's only the compression type and level which are passed, and in this particular case the memory pool. However, not all three are always passed but rather the compression level and codec. Thus, I do not want to pass properties since there is a lot more unnecessary things there.

wesm commented on cpp/src/parquet/file_writer.cc:
```diff
@@ -221,10 +241,13 @@ class RowGroupSerializer : public RowGroupWriter::Contents {
-          PageWriter::Open(sink_, properties_->compression(column_descr->path()),
-                           col_meta, properties_->memory_pool(), buffered_row_group_);
+          PageWriter::Open(sink_, compressionCodec, compression_level, col_meta,
+                           properties_->memory_pool(), buffered_row_group_);
```
Same comments as above re: this code duplication

martinradev replied:
Answered  above.

wesm commented on cpp/src/parquet/types.cc:
```diff
@@ -34,32 +34,44 @@ using arrow::util::Codec;
+      TranslateParquetCompressionTypeToArrowCompressionType(codec);
+  PARQUET_THROW_NOT_OK(Codec::Create(compression_type, &result));
+  return result;
+}
```
It seems like this function is not needed anymore?

martinradev replied:
Removed

wesm commented on cpp/src/parquet/types.cc:
```diff
@@ -34,32 +34,44 @@ using arrow::util::Codec;
-std::unique_ptr<Codec> GetCodecFromArrow(Compression::type codec) {
-  std::unique_ptr<Codec> result;
+namespace {
+arrow::Compression::type TranslateParquetCompressionTypeToArrowCompressionType(
```
Seems silly to me to continue to maintain this code. Can we do `using Compression = ::arrow::Compression` in `parquet/types.h` and move on?

martinradev replied:
By this, you mean parquet::Compression to be removed and replaced with arrow::Compression?
The API would be broken then since this would also expose Compression::BZ2 which is not a valid compression algorithm for Parquet.
I do not know the history of the project and why this is so.

wesm replied:
I think we would prohibit unsupported compressions in Parquet. 

@xhochy @majetideepak any thoughts about this?

xhochy replied:
I'm fine with this either way as long as we are to get programmatically which codecs are supported by Parquet.

wesm commented on cpp/src/parquet/types.cc:
```diff
@@ -259,6 +271,11 @@ std::string ConvertedTypeToString(ConvertedType::type t) {
+int32_t GetCompressionCodecDefaultCompressionLevel(Compression::type t) {
+  return arrow::util::GetCompressionCodecDefaultCompressionLevel(
+      TranslateParquetCompressionTypeToArrowCompressionType(t));
+}
```
If my suggested change is made this is not needed

martinradev replied:
Removed.

wesm commented:
@martinradev the PR description is no longer accurate, can you update it for the sake of the changelog?

martinradev commented on cpp/src/arrow/util/compression.cc:
```diff
@@ -48,13 +48,42 @@
 namespace arrow {
 namespace util {
 
+int32_t GetCompressionCodecDefaultCompressionLevel(Compression::type t) {
```
I think it doesn't because then this would add a dependency on the implementation of the codec. The implementation would only be visible if the library is compiled in and this would introduce a lot of ifdefs (ifdef ARROW_WITH_ZLIB, etc).

However, I can change it as you requested. I'm probably not getting this patch in if I would be stubborn for everything.

martinradev replied:
I decided to follow what @pitrou recommended by using numeric_limits<int>::min as a hint to let the implementation select whatever compression level it deems appropriate.

wesm replied:
I'm making this a static method... I will push changes to your branch

martinradev commented on cpp/src/arrow/util/compression_bz2.cc:
```diff
@@ -37,8 +37,6 @@
 namespace arrow {
 namespace util {
 
-constexpr int kBZ2DefaultCompressionLevel = 9;
```
Undid my change with regards to this.

martinradev commented:
@wesm @pitrou Thanks for the reviews. I made some more changes and I think the current version of the patch is closer to what @pitrou wanted from the start.
I also changed the type of the compression level from int32_t to int since this is what all supported compressors with the option for level selection expect.

Can you give another review?

martinradev commented on cpp/src/parquet/properties.h:
```diff
@@ -335,6 +399,7 @@ class PARQUET_EXPORT WriterProperties {
     ColumnProperties default_column_properties_;
     std::unordered_map<std::string, Encoding::type> encodings_;
     std::unordered_map<std::string, Compression::type> codecs_;
+    std::unordered_map<std::string, int32_t> codecs_compression_level_;
```
Ok, this should be int. I can write a follow-up patch if there are no more things to fix in this one.
If there are more things to fix, then I will fix it alongside whatever needs to be addressed.

martinradev commented:
@wesm Can you have a second look?

wesm commented:
Looking. In the future, can you use a feature branch? Using master makes things more difficult for project maintainers

wesm commented on cpp/src/parquet/properties.h:
```diff
@@ -271,6 +280,59 @@ class PARQUET_EXPORT WriterProperties {
+     * If no level is selected by the user or if the special
+     * std::numeric_limits<int>::min() value is passed, then Arrow selects the compression
+     * level.
+     */
```
We use `///` comments. I will fix

wesm commented:
Bah, I got thwarted by the GitHub UI. It's difficult for me to edit your branch when it's "master" so I will open a new PR

wesm commented:
See https://github.com/apache/arrow/pull/5192
